api/include/opentelemetry/logs/logger_provider.h

```diff
@@ -44,4 +45,29 @@ class LoggerProvider
 };
 }  // namespace logs
 OPENTELEMETRY_END_NAMESPACE
-#endif
+
+#  if defined(OPENTELEMETRY_EXPORT)
+
+namespace std
+{
+
+//
+// Partial specialization of default_delete used by unique_ptr.
+// This makes the delete of the type in unique_ptr happening in the DLL where it
+// is allocated.
+//
+template <>
+class OPENTELEMETRY_EXPORT default_delete<OPENTELEMETRY_NAMESPACE::logs::LoggerProvider>
```

Comment:
The default_delete is the destruction policy used by unique_ptr, however the LoggerProvider is of type nostd::shared_ptr. How will this work with shared_ptr.

Reply:
Yes, this the shared_ptr could be a problem here. Thanks for the catch, @marcalff and @lalitb . While I am looking more into it, is there any recommendation on preventing the memory free pattern?

Reply:
According to [cpp_reference](https://en.cppreference.com/w/cpp/memory/unique_ptr), for unique pointers, the default deleter is:

```
class Deleter = std::default_delete<T>
```

For shared_ptr, the [doc](https://en.cppreference.com/w/cpp/memory/shared_ptr/shared_ptr) does not provide a default, and gives constraints for the expected behavior of the deleter:

```
template< class Y, class Deleter >
shared_ptr( Y* ptr, Deleter d );
```

If I understand correctly, it means we will have to implement Deleter classes for every shared pointer used,
and declare shared_ptr as:

```
shared_ptr<TracerProvider, TracerProviderDeleter>
```

In this case, implementing the deleter twice (once as std::default_delete<> for unique_ptr, once for shared_ptr, is an issue. We may use TracerProviderDeleter as well for unique_ptr and never rely on the default.

Also, in this case, we need typedefs to make the code more readable:

typedef shared_ptr<TracerProvider, TracerProviderDeleter> TracerProviderSharedPtr;
typedef unique_ptr<TracerProvider, TracerProviderDeleter> TracerProviderUniquePtr;

or maybe helper types:

TracerProvider::shared_ptr
TracerProvider::unique_ptr

If this turns out to be required, this change is going to break existing code and abi.

Reply:
Here is the change for this.
```diff
  explicit shared_ptr(pointer ptr)
  {
-    std::shared_ptr<T> ptr_(ptr);
+    std::shared_ptr<T> ptr_(ptr, std::default_delete<T>());
```

Reply:
This will fix only nostd::shared_ptr<T> when building without C++ 11 std::shared_ptr, WITH_STL=OFF.

What about the build where nostd::shared_ptr is an alias for std::shared_ptr, WITH_STL=ON ?

This code will not be used, and the deleter for std::shared_ptr<T> will be the same as before.

Reply:
This is a good point. To fix `WITH_STIL=ON`, we may need implement the 5th constructor of `shared_ptr` for `nostd::shared_ptr`, then passing the deleter in the caller (like in [GetLogger](https://github.com/open-telemetry/opentelemetry-cpp/blob/1ae2fe538ceaaf3829c2972a05e043cf43676000/sdk/src/logs/logger_provider.cc#L110)). 

I'd prefer emitting an error for DLL build when `WITH_STL=ON` and fix this later.

Reply:
As the major purpose of providing nostd class is for ABI compatibility, I am wondering whether we recommend DLL user to build with `WITH_STIL=ON`, or we should not recommend that because that doesn't provide ABI compatibility.

Reply:
Added error out for DLL build and `WITH_STL=ON`.